server/bin/pbench-audit-server.py

```diff
@@ -0,0 +1,120 @@
+#! /usr/bin/env python3
+
+import os
+import sys
+import glob
+
+from pbench import PbenchConfig, BadConfig, get_pbench_logger
+
+_NAME_ = 'pbench-audit-server'
+
+class Hierarchy(object):
+
+    def __init__(self, name, path, config):
+        self.name = name
+        self.path = path
+        self.config = config
+
+    def dump(self, n, msg):
+        print("\t*",msg)
+        print("\t  ++++++++++", end='\n\t  ')
+        print(*n, sep='\n\t  ')
+        print("\t  ----------")
+
+
+def verify_subdirs(hier, controller):
+    cnt=0
+    linkdirs = sorted(hier.config.LINKDIRS.split(' '))
+
+    direc_entries = glob.iglob(os.path.join(controller,'*'))
+    tarball, symlink, unexpected_objects, unexpected_dirs, expected_dirs = (list() for i in range(5))
+
+    for dirents in direc_entries:
+        st = os.path.basename(os.path.normpath(dirents))
+        if os.path.isdir(dirents):
+            expected_dirs.append(st)
+            if st != '_QUARANTINED' and not st.startswith('WONT-INDEX'):
+                if st not in linkdirs:
+                    unexpected_dirs.append(st)
+            continue
+        if os.path.isfile(dirents):
+            if st.endswith(('.tar.xz', '.tar.xz.md5')):
+                tarball.append(st)
+            else:
+                unexpected_objects.append(st)
+            continue
+        if os.path.islink(dirents):
+            symlink.append(st)
+            continue
+
+    if unexpected_dirs or not expected_dirs or not tarball or symlink or unexpected_objects:
+        print("\nController:", os.path.basename(controller))
+        if unexpected_dirs:
+            hier.dump(unexpected_dirs, "Unexpected state directories found in this controller directory:")
+            cnt = cnt+1
+        if not expected_dirs:
+            print("\t* No state directories found in this controller directory.")
+        if not tarball:
+            print("\t* No tar ball files found in this controller directory.")
+            cnt=cnt+1
+        if symlink:
+            hier.dump(symlink, "Unexpected symlinks in controller directory:")
+            cnt = cnt+1
+        if unexpected_objects:
+            hier.dump(unexpected_objects, "Unexpected files in controller directory:")
+            cnt = cnt+1
+
+    return cnt
+
+def verify_prefixes(hier, controller):
+    pass
+
+def verify_archive(hier):
+    cnt=0
+    print("\nstart-{}: archive hierarchy: {}".format(hier.config.TS[4:], hier.config.ARCHIVE))
+    controllers = glob.iglob(os.path.join(hier.path, '*'))
+
+    bad_controller = list()
```

Comment:
This should be a class attribute:
       hier.bad_controllers = list()

Reply:
yaa got it